File: ML/schemas/action_proposal.py
"""ActionProposal schema — mirrors CONTRACTS.md §13/§14.

The ActionProposal is the boundary between Agent Intelligence and SentinelPay
security/execution. It describes what the agent WANTS to do. It is NOT a signed
transaction, NOT authorization, and must never contain a private key
(CONTRACTS.md §14).
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

ActionType = Literal["PAY", "TRANSFER", "PURCHASE", "CALL_API", "SWAP", "OTHER"]


class Money(BaseModel):
    """Decimal-string amount, never float (CONTRACTS.md §4.3)."""

    value: str
    currency: str


class RecipientRef(BaseModel):
    type: Literal["SERVICE", "WALLET", "CONTRACT", "MERCHANT", "UNKNOWN"] = "UNKNOWN"
    identifier: str | None = None
    address: str | None = None
    network: str | None = None


class ActionProposal(BaseModel):
    schema_version: Literal["action_proposal.v1"] = "action_proposal.v1"
    proposal_id: str
    intent_id: str
    agent_id: str

    action_type: ActionType
    purpose: str

    recipient: RecipientRef | None = None
    amount: Money | None = None
    payment_method: str | None = None
    network: str | None = None

    reason: str = ""
    evidence_ids: list[str] = Field(default_factory=list)
    trajectory_event_ids: list[str] = Field(default_factory=list)

    created_at: str

    # Explicit guard: this object must never carry authority/execution fields.
    # (Enforced structurally by NOT declaring them; a test asserts their absence.)

File: Backend/src/shared/contracts/action-proposal.ts
/**
 * ActionProposal contract — Phase 2, Task 2.1.
 *
 * Mirrors CONTRACTS.md §13/§14 and Orvex/ML/schemas/action_proposal.py. The
 * ActionProposal is the boundary between Agent Intelligence and SentinelPay
 * security/execution: it describes what the agent WANTS to do. It is NOT a signed
 * transaction, NOT authorization, and MUST NEVER contain a private key
 * (CONTRACTS.md §14). The absence of authority/key fields is enforced structurally
 * (the fields are simply not declared) and asserted by the authority-boundary tests.
 *
 * Consumer-side: `.passthrough()` for forward-compat.
 */
import { z } from "zod";
import { idString, isoTimestamp, Money } from "./primitives.js";

export const ActionType = z.enum([
  "PAY",
  "TRANSFER",
  "PURCHASE",
  "CALL_API",
  "SWAP",
  "OTHER",
]);
export type ActionType = z.infer<typeof ActionType>;

export const RecipientRefType = z.enum([
  "SERVICE",
  "WALLET",
  "CONTRACT",
  "MERCHANT",
  "UNKNOWN",
]);
export type RecipientRefType = z.infer<typeof RecipientRefType>;

export const RecipientRef = z
  .object({
    type: RecipientRefType.default("UNKNOWN"),
    identifier: z.string().nullish(),
    address: z.string().nullish(),
    network: z.string().nullish(),
  })
  .passthrough();
export type RecipientRef = z.infer<typeof RecipientRef>;

export const ActionProposal = z
  .object({
    schema_version: z
      .literal("action_proposal.v1")
      .default("action_proposal.v1"),
    proposal_id: idString,
    intent_id: idString,
    agent_id: idString,

    action_type: ActionType,
    purpose: z.string(),

    recipient: RecipientRef.nullish(),
    amount: Money.nullish(),
    payment_method: z.string().nullish(),
    network: z.string().nullish(),

    reason: z.string().default(""),
    evidence_ids: z.array(z.string()).default([]),
    trajectory_event_ids: z.array(z.string()).default([]),

    created_at: isoTimestamp,
  })
  .passthrough();
export type ActionProposal = z.infer<typeof ActionProposal>;
